File: P_2/Examen_P2/Napakalaki_Ruby_P2/lib/player.rb
#encoding: utf-8
# MADE BY: JOSE MIGUEL HDEZ GARCIA & JAIRO ABRIL MOYA

require_relative 'treasure'
require_relative 'monster'
require_relative 'treasure_kind'
require_relative 'combat_result'
require_relative 'dice'
require_relative 'napakalaki'

class Player

    # Consultores basicos
    attr_reader :name
    attr_reader :level
    attr_reader :dead
    attr_reader :canISteal
    attr_reader :pending_badconsequence
    attr_reader :visible_treasures
    attr_reader :hidden_treasures
    attr_writer :enemy

    def initialize(name)
        @name = name
        @level = 1
        @dead = true
        @canISteal = true
        @visible_treasures = Array.new
        @hidden_treasures = Array.new
    end

    def is_dead
        return @dead
    end

    def get_combat_level
        @lvl = level
        return @lvl
    end

    def bring_to_life
        @dead = false
    end

    def increment_levels(l)
        @level += l
    end

    def decrement_levels(l)
        @level -= l

        if (@level < 1)
            @level = 1
        end
    end

    def set_pending_badconsequence(bc)
        @pending_badconsequence = bc
    end

    def apply_prize(current_monster)

    end

    def apply_badconsequence(bc)

    end

    def can_make_treasure_visible(t)

    end

    def how_many_visible_treasures(t_kind)

        i = 0
        @visible_treasures.each do |t|

            if t.type == t_kind then
                i = i + 1
            end
        end
        return i
    end

    def die_if_no_treasures
        if @visible_treasures.empty? && @hidden_treasures.empty?
            @dead = true
        end
    end

    def valid_state
       return @pending_badConsequence != nil && @hidden_treasures.length <= 5
    end

    def can_i_steal
        @canISteal = true
    end

    def give_me_a_treasure
        random = rand(@hidden_treasures.length)
        return @hidden_treasures.remove(random)
    end

    def canYouGiveMeATreasure?
        return ( !(@visibleTreasures.empty?) && !(@hiddenTreasures.empty?))
    end


end

File: P_2/Examen_P2/Napakalaki_Java_P2/src/napakalaki/Player.java
// MADE BY: JOSE MIGUEL HDEZ GARCIA & JAIRO ABRIL MOYA

package napakalaki;
import java.util.ArrayList;
import java.util.Random;

/**
 *
 * @author ixjosemi
 */

public class Player {

    static int MAXLEVEL = 10;
    private String name;
    private int level;
    private boolean dead = true;
    private boolean canISteal = true;
    private BadConsequence pendingBadConsequence;
    private Player enemy;

    private ArrayList <Treasure> visibleTreasures = new ArrayList();
    private ArrayList <Treasure> hiddenTreasures = new ArrayList();

    public Player(String name){
        this.name = name;
        this.level = 1;
    }

    public Player (Player p){
        this.name = name;
        this.level = p.level;
        this.pendingBadConsequence = p.pendingBadConsequence;
        this.dead = p.dead;
        this.canISteal = p.canISteal;
    }

    public String getName(){
        return name;
    }

    private void bringToLife(){
        dead = false;
    }

    private int getCombatLevel(){
        int lvl = this.level;
        return lvl;
    }

    private void incrementLevels(int levels){
        this.level += levels;

    }

    private void decrementeLevels(int levels){
        this.level -= levels;

        if (this.level <= 1)
            this.level = 1;
    }

    private void setPendingBadConsequence(BadConsequence bc){
        this.pendingBadConsequence = bc;
    }

    private void applyPrize(Monster m){

    }

    private void applyBadConsequence(Monster m){

    }

    /*
    private boolean canMakeTreasureVisible(Treasure t){

    }
    */

    private int howManyVisibleTreasures(TreasureKind tKind){
        int i = 0;

        for(Treasure t  : this.visibleTreasures){

            if (t.getType() == tKind)
                i++;
        }
        return i;
    }

    private void dieIfNoTreasure(){
        if(this.visibleTreasures.isEmpty() && this.hiddenTreasures.isEmpty())
            this.dead = true;
    }

    BadConsequence bc;

    public boolean isDead(){
        dead = bc.getDeath();
        return dead;
    }

    public ArrayList getHiddenTreasures(){
        return hiddenTreasures;
    }

    public ArrayList getVisibleTreasure(){
        return visibleTreasures;
    }

    /*
    public CombatResult combat(Monster m){

    }
    */

    public void makeTreasureVisible(Treasure t){

    }

    public void discardVisibleTreasure(Treasure t){

    }

    public void discardHiddenTreasure(Treasure t){

    }

    public boolean validState(){
        return this.pendingBadConsequence == null
                || (this.pendingBadConsequence.isEmpty() && this.hiddenTreasures.size() <= 5);
    }

    public void initTreasure(){

    }

    public int getLevels(){
        return this.level;
    }

    /*
    public Treasure stealTreasure(){

    }
    */

    public void setEnemy(Player enemy){
        this.enemy = enemy;
    }

    private Treasure giveMeATreasure(){
        Random rnd = new Random();

        int i = rnd.nextInt(hiddenTreasures.size());

        return(hiddenTreasures.remove(i));
    }

    public boolean canISteal(){
        return this.canISteal;
    }

    private boolean canYouGiveMeATreasure(){
        boolean have_treasures = true;

        if (hiddenTreasures.isEmpty() && visibleTreasures.isEmpty())
            have_treasures = false;

        return (have_treasures);
    }

    private void haveStolen(){
        canISteal = false;
    }

    public void discardAllTreasures(){

    }
}
